pkg/controller/pkg/revision/dependency.go

```diff
@@ -0,0 +1,231 @@
+		}
+		pdep.Constraints = dep.Version
+		sources[i] = pdep
+	}
+
+	total = len(sources)
+
+	// Get the lock.
+	lock := &v1alpha1.Lock{}
+	if err := m.client.Get(ctx, types.NamespacedName{Name: lockName}, lock); err != nil {
+		return total, installed, invalid, errors.Wrap(err, errGetLock)
+	}
+
+	prRef, err := name.ParseReference(pr.GetSource())
+	if err != nil {
+		return total, installed, invalid, err
+	}
+
+	selfIndex := intPointer(-1)
+	d := m.newDag()
+	implied, err := d.Init(v1alpha1.ToNodes(lock.Packages), dag.FindIndex(prRef.Context().String(), selfIndex))
+	if err != nil {
+		return total, installed, invalid, err
+	}
+
+	// If we are inactive, all we want to do is remove self.
+	if pr.GetDesiredState() == v1beta1.PackageRevisionInactive {
+		if *selfIndex >= 0 {
+			lock.Packages = append(lock.Packages[:*selfIndex], lock.Packages[*selfIndex+1:]...)
+			return total, installed, invalid, m.client.Update(ctx, lock)
+		}
+		return total, installed, invalid, nil
+	}
+
+	// NOTE(hasheddan): consider adding health of package to lock so that it can
+	// be rolled up to any dependent packages.
+	self := v1alpha1.LockPackage{
+		Name: pr.GetName(),
+		Type: m.packageType,
+		// NOTE(hasheddan): go-containerregistry appends default registry if
+		// none is provided. Make sure to use the same method to add a package
+		// to the lock as is used to identify its index above.
+		Source:       prRef.Context().String(),
+		Version:      prRef.Identifier(),
+		Dependencies: sources,
+	}
+
+	// If we don't exist in lock then we should add self.
+	if *selfIndex == -1 {
+		lock.Packages = append(lock.Packages, self)
+		if err := m.client.Update(ctx, lock); err != nil {
+			return total, installed, invalid, err
+		}
+		// Package may exist in the graph as a dependency, or may not exist at
+		// all. We need to either convert it to a full node or add it.
+		d.AddOrUpdateNodes(&self)
+
+		// If any direct dependencies are missing we skip checking for
+		// transitive ones.
+		var missing []string
+		for _, dep := range self.Dependencies {
+			if d.NodeExists(dep.Identifier()) {
+				installed++
+				continue
+			}
+			missing = append(missing, dep.Identifier())
+		}
+		if installed != total {
+			return total, installed, invalid, errors.Errorf(errMissingDependenciesFmt, missing)
+		}
+	}
+
+	tree := map[string]dag.Node{}
+	if err := d.TraceNode(prRef.Context().String(), tree); err != nil {
+		return total, installed, invalid, err
+	}
+	total = len(tree)
+	installed = total
+	// Check if any dependencies or transitive dependencies are missing (implied).
+	var missing []string
+	for _, imp := range implied {
+		if _, ok := tree[imp.Identifier()]; ok {
+			installed--
+			missing = append(missing, imp.Identifier())
+		}
+	}
+	if len(missing) != 0 {
+		return total, installed, invalid, errors.Errorf(errMissingDependenciesFmt, missing)
+	}
+
+	// All of our dependencies and transitive dependencies must exist. Check
+	// that neighbors have valid versions.
```

Comment:
Note that we do not gate on dependencies actually being healthy. As long as they exist we continue with installation. It could be an improvement to check dependency health in the future, but right now it would likely cause unneeded additional reconciles. This could be improved if we are able to stop reconciling every minute for revisions.